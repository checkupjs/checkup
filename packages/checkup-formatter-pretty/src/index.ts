import * as React from 'react';
import { render } from 'ink-render-string';
import {
  CheckupLogParser,
  Formatter,
  FormatterOptions,
  ErrorKind,
  CheckupError,
} from '@checkup/core';
import { registerCustomComponent } from './component-provider';
import { default as pretty } from './pretty-formatter';

class PrettyFormatter implements Formatter {
  options: FormatterOptions;
  component: any;

  constructor(options: FormatterOptions) {
    this.options = options;
  }

  format(logParser: CheckupLogParser): string {
    try {
      const result = render(React.createElement(pretty, { logParser }));

      if (result.includes('ERROR')) {
        throw result;
      } else {
        return result;
      }
    } catch (error) {
      throw new CheckupError(ErrorKind.InvalidCustomComponent, {
        details: error,
      });
    }
  }

  static registerCustomComponent = registerCustomComponent;
}

// @ts-ignore
export = PrettyFormatter;
