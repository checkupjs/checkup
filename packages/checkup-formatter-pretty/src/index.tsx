import { Formatter, FormatterArgs } from '@checkup/core';
import * as React from 'react';
import { render } from 'ink';

import Pretty from './pretty';

class PrettyFormatter implements Formatter {
  constructor(public args: FormatterArgs) {}

  async format() {
    let data: string[] = ['table', 'list'];

    let app = render(<Pretty data={data} />);

    await app.waitUntilExit();
  }
}

export = PrettyFormatter;
