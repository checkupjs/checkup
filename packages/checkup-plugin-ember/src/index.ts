import { RegistrationArgs, getPluginName } from '@checkup/core';

import { evaluateActions as evaluateTemplateLintDisables } from './actions/ember-template-lint-disable-actions';
import { evaluateActions as evaluateTemplateLintSummary } from './actions/ember-template-lint-summary-actions';
import { evaluateActions as evaluateTestTypes } from './actions/ember-test-types-actions';
import { report as reportEmberTestTypes } from './reporters/ember-test-types-reporter';
import { report as reportEmberOctaneMigrationStatus } from './reporters/ember-octane-migration-status-reporter';
import { report as reportEmberTemplateLintSummary } from './reporters/ember-template-lint-summary-reporter';
import EmberDependenciesTask from './tasks/ember-dependencies-task';
import EmberInRepoAddonsEnginesTask from './tasks/ember-in-repo-addons-engines-task';
import EmberTestTypesTaskTask from './tasks/ember-test-types-task';
import EmberTypesTask from './tasks/ember-types-task';
import EmberTemplateLintDisableTask from './tasks/ember-template-lint-disable-task';
import EmberTemplateLintSummaryTask from './tasks/ember-template-lint-summary-task';
import EmberOctaneMigrationStatusTask from './tasks/ember-octane-migration-status-task';

export function register(args: RegistrationArgs) {
  let pluginName = getPluginName(__dirname);

  args.register.actions('ember-template-lint-disables', evaluateTemplateLintDisables);
  args.register.actions('ember-template-lint-summary', evaluateTemplateLintSummary);
  args.register.actions('ember-test-types', evaluateTestTypes);

  args.register.taskReporter('ember-test-types', reportEmberTestTypes);
  args.register.taskReporter('ember-octane-migration-status', reportEmberOctaneMigrationStatus);
  args.register.taskReporter('ember-template-lint-summary', reportEmberTemplateLintSummary);

  args.register.task(new EmberTypesTask(pluginName, args.context));
  args.register.task(new EmberDependenciesTask(pluginName, args.context));
  args.register.task(new EmberInRepoAddonsEnginesTask(pluginName, args.context));
  args.register.task(new EmberTestTypesTaskTask(pluginName, args.context));
  args.register.task(new EmberTemplateLintDisableTask(pluginName, args.context));
  args.register.task(new EmberTemplateLintSummaryTask(pluginName, args.context));
  args.register.task(new EmberOctaneMigrationStatusTask(pluginName, args.context));
}
