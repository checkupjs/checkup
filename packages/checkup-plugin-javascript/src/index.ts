import { RegistrationArgs, getPluginName } from '@checkup/core';

import { evaluateActions as evaluateESLintDisables } from './actions/eslint-disable-actions';
import { evaluateActions as evaluateESLintSummary } from './actions/eslint-summary-actions';
import { evaluateActions as evaluateOutdatedDependencies } from './actions/outdated-dependency-actions';
import EslintDisableTask from './tasks/eslint-disable-task';
import OutdatedDependencyTask from './tasks/outdated-dependencies-task';
import EslintSummaryTask from './tasks/eslint-summary-task';

export function register(args: RegistrationArgs) {
  let pluginName = getPluginName(__dirname);

  args.register.actions('eslint-disables', evaluateESLintDisables);
  args.register.actions('eslint-summary', evaluateESLintSummary);
  args.register.actions('outdated-dependencies', evaluateOutdatedDependencies);

  args.register.task(new EslintSummaryTask(pluginName, args.context));
  args.register.task(new EslintDisableTask(pluginName, args.context));
  args.register.task(new OutdatedDependencyTask(pluginName, args.context));
}
