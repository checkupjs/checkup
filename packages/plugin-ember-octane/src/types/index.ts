import { CLIEngine } from 'eslint';
import { TemplateLintResult } from './ember-template-lint';

export enum ESLintMigrationType {
  NativeClasses = 'native-classes',
  TaglessComponents = 'tagless-components',
  GlimmerComponents = 'glimmer-components',
  TrackedProperties = 'tracked-properties',
}

export enum TemplateLintMigrationType {
  AngleBrackets = 'angle-brackets',
  NamedArgs = 'named-args',
  OwnProperties = 'own-properties',
  UseModifiers = 'use-modifiers',
}

type LintResultCollection = CLIEngine.LintResult[] | TemplateLintResult[];

interface CompletionInfo {
  total: number;
  completed: number;
  percentage: string;
}

export interface MigrationInfo {
  completionInfo: CompletionInfo;
  name: string;
  relatedResults: LintResultCollection;
}

export interface MigrationRuleConfig {
  fileMatchers: RegExp[];
  name: string;
  rules: string[];
}
