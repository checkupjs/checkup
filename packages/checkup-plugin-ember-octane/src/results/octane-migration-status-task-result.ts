import {
  BaseTaskResult,
  ESLintReport,
  PieChartData,
  TaskMetaData,
  TaskResult,
  TemplateLintReport,
  ui,
} from '@checkup/core';
import {
  ESLINT_MIGRATION_TASK_CONFIGS,
  TEMPLATE_LINT_MIGRATION_TASK_CONFIGS,
} from '../utils/task-configs';
import { ESLintMigrationType, MigrationInfo, TemplateLintMigrationType } from '../types';
import { transformESLintReport, transformTemplateLintReport } from '../utils/transformers';

export default class OctaneMigrationStatusTaskResult extends BaseTaskResult implements TaskResult {
  migrationResults: MigrationInfo[];
  totalViolations: number;

  constructor(
    meta: TaskMetaData,
    public esLintReport: ESLintReport,
    public templateLintReport: TemplateLintReport
  ) {
    super(meta);
    this.migrationResults = this.formattedMigrationResults;
    this.totalViolations = this.esLintReport.errorCount + this.templateLintReport.errorCount;
  }

  stdout() {
    ui.section(this.meta.friendlyTaskName, () => {
      ui.log(`${ui.emphasize('Octane Violations')}: ${this.totalViolations}`);
      ui.blankLine();
      ui.table(this.migrationResults, {
        name: { header: 'Octane Feature' },
        completion: {
          header: 'Completion',
          get: (row: MigrationInfo) => `${row.completionInfo.percentage}%`,
        },
      });
    });
  }

  json() {
    return {
      meta: this.meta,
      result: {
        totalViolations: this.esLintReport.errorCount + this.templateLintReport.errorCount,
        migrationTaskResults: this.migrationResults,
      },
    };
  }

  html() {
    return this.migrationResults
      .map((migrationResult) => this._createPieChartData(migrationResult))
      .filter(Boolean) as PieChartData[];
  }

  _createPieChartData(migrationResult: MigrationInfo): PieChartData | undefined {
    if (Object.keys(migrationResult).length === 0) {
      return;
    }

    return new PieChartData(
      this.meta,
      [
        { value: migrationResult.completionInfo.completed, description: 'migrated' },
        {
          value: migrationResult.completionInfo.total - migrationResult.completionInfo.completed,
          description: 'unmigrated',
        },
      ],
      migrationResult.name
    );
  }

  get formattedMigrationResults() {
    let eslintMigrationTasks: ESLintMigrationType[] = [
      ESLintMigrationType.NativeClasses,
      ESLintMigrationType.TaglessComponents,
      ESLintMigrationType.GlimmerComponents,
      ESLintMigrationType.TrackedProperties,
    ];
    let templateLintMigrationTasks: TemplateLintMigrationType[] = [
      TemplateLintMigrationType.AngleBrackets,
      TemplateLintMigrationType.NamedArgs,
      TemplateLintMigrationType.OwnProperties,
      TemplateLintMigrationType.UseModifiers,
    ];

    let eslintMigrationResults = eslintMigrationTasks.map((eslintMigrationTask) =>
      transformESLintReport(ESLINT_MIGRATION_TASK_CONFIGS[eslintMigrationTask], this.esLintReport)
    );

    let templateLintMigrationResults = templateLintMigrationTasks.map((templateMigrationTask) =>
      transformTemplateLintReport(
        TEMPLATE_LINT_MIGRATION_TASK_CONFIGS[templateMigrationTask],
        this.templateLintReport
      )
    );
    return [...eslintMigrationResults, ...templateLintMigrationResults];
  }
}
