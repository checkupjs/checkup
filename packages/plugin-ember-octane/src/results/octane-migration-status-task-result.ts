import { CLIEngine } from 'eslint';
import { BaseTaskResult, TaskMetaData, TaskResult, ui } from '@checkup/core';
import { ESLintMigrationType, MigrationInfo, TemplateLintMigrationType } from '../types';
import { TemplateLintReport } from '../types/ember-template-lint';
import { transformESLintReport, transformTemplateLintReport } from '../utils/transformers';
import {
  ESLINT_MIGRATION_TASK_CONFIGS,
  TEMPLATE_LINT_MIGRATION_TASK_CONFIGS,
} from '../utils/task-configs';

export default class OctaneMigrationStatusTaskResult extends BaseTaskResult implements TaskResult {
  taskName: string = 'Octane Migration Status';

  constructor(
    meta: TaskMetaData,
    public esLintReport: CLIEngine.LintReport,
    public templateLintReport: TemplateLintReport
  ) {
    super(meta);
  }

  stdout() {
    let jsonOutput = this.json();
    let { esLint: esLintResults, templateLint: templateLintResults } = jsonOutput.result;

    let migrationTasks = [
      esLintResults.migrationTasks[ESLintMigrationType.NativeClasses],
      esLintResults.migrationTasks[ESLintMigrationType.TaglessComponents],
      esLintResults.migrationTasks[ESLintMigrationType.GlimmerComponents],
      esLintResults.migrationTasks[ESLintMigrationType.TrackedProperties],
      templateLintResults.migrationTasks[TemplateLintMigrationType.AngleBrackets],
      templateLintResults.migrationTasks[TemplateLintMigrationType.NamedArgs],
      templateLintResults.migrationTasks[TemplateLintMigrationType.OwnProperties],
      templateLintResults.migrationTasks[TemplateLintMigrationType.UseModifiers],
    ];

    ui.styledHeader(this.taskName);
    ui.blankLine();
    ui.styledObject({
      'JavaScript Octane Violations': esLintResults.totalViolations,
      'Handlebars Octane Violations': templateLintResults.totalViolations,
    });
    ui.blankLine();
    ui.table(migrationTasks, {
      name: { header: 'Migration Task' },
      completion: {
        header: 'Completion Percentage',
        get: (row: MigrationInfo) => `${row.completionInfo.percentage}%`,
      },
    });
    ui.blankLine();
  }

  json() {
    let nativeClassMigrationInfo = transformESLintReport(
      ESLINT_MIGRATION_TASK_CONFIGS[ESLintMigrationType.NativeClasses],
      this.esLintReport
    );

    let taglessComponentMigrationInfo = transformESLintReport(
      ESLINT_MIGRATION_TASK_CONFIGS[ESLintMigrationType.TaglessComponents],
      this.esLintReport
    );

    let glimmerComponentsMigrationinfo = transformESLintReport(
      ESLINT_MIGRATION_TASK_CONFIGS[ESLintMigrationType.GlimmerComponents],
      this.esLintReport
    );

    let trackedPropertiesMigrationInfo = transformESLintReport(
      ESLINT_MIGRATION_TASK_CONFIGS[ESLintMigrationType.TrackedProperties],
      this.esLintReport
    );

    let angleBracketsMigrationInfo = transformTemplateLintReport(
      TEMPLATE_LINT_MIGRATION_TASK_CONFIGS[TemplateLintMigrationType.AngleBrackets],
      this.templateLintReport
    );

    let namedArgsMigrationInfo = transformTemplateLintReport(
      TEMPLATE_LINT_MIGRATION_TASK_CONFIGS[TemplateLintMigrationType.NamedArgs],
      this.templateLintReport
    );

    let ownPropsMigrationInfo = transformTemplateLintReport(
      TEMPLATE_LINT_MIGRATION_TASK_CONFIGS[TemplateLintMigrationType.OwnProperties],
      this.templateLintReport
    );

    let useModifiersMigrationInfo = transformTemplateLintReport(
      TEMPLATE_LINT_MIGRATION_TASK_CONFIGS[TemplateLintMigrationType.UseModifiers],
      this.templateLintReport
    );

    return {
      meta: this.meta,
      result: {
        esLint: {
          totalViolations: this.esLintReport.errorCount,
          migrationTasks: {
            [ESLintMigrationType.NativeClasses]: nativeClassMigrationInfo,
            [ESLintMigrationType.TaglessComponents]: taglessComponentMigrationInfo,
            [ESLintMigrationType.GlimmerComponents]: glimmerComponentsMigrationinfo,
            [ESLintMigrationType.TrackedProperties]: trackedPropertiesMigrationInfo,
          },
        },
        templateLint: {
          totalViolations: this.templateLintReport.errorCount,
          migrationTasks: {
            [TemplateLintMigrationType.AngleBrackets]: angleBracketsMigrationInfo,
            [TemplateLintMigrationType.NamedArgs]: namedArgsMigrationInfo,
            [TemplateLintMigrationType.OwnProperties]: ownPropsMigrationInfo,
            [TemplateLintMigrationType.UseModifiers]: useModifiersMigrationInfo,
          },
        },
      },
    };
  }

  pdf() {
    return undefined;
  }
}
