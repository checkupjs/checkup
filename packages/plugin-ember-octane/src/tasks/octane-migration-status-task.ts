import { BaseTask, Category, Priority, Task } from '@checkup/core';

import { CLIEngine } from 'eslint';
import { OctaneMigrationStatusTaskResult } from '../results';

export default class OctaneMigrationStatusTask extends BaseTask implements Task {
  meta = {
    taskName: 'octane-migration-status',
    friendlyTaskName: 'Ember Octane Migration Status',
    taskClassification: {
      category: Category.Core,
      priority: Priority.Medium,
    },
  };

  public report!: CLIEngine.LintReport;
  private esLintEngine: CLIEngine;

  constructor(cliArguments: any) {
    super(cliArguments);

    // These options are taken from a default ember application build on top of ember-source 3.16.*
    this.esLintEngine = new CLIEngine({
      parser: 'babel-eslint',
      parserOptions: {
        ecmaVersion: 2018,
        sourceType: 'module',
        ecmaFeatures: {
          legacyDecorators: true,
        },
      },
      plugins: ['ember'],
      envs: ['browser'],
      rules: {
        'ember/classic-decorator-hooks': 'error',
        'ember/classic-decorator-no-classic-methods': 'error',
        'ember/no-actions-hash': 'error',
        'ember/no-classic-classes': 'error',
        'ember/no-classic-components': 'error',
        // 'ember/no-component-lifecycle-hooks': 'error',
        'ember/no-computed-properties-in-native-classes': 'error',
        'ember/no-get-with-default': 'error',
        'ember/no-get': 'error',
        'ember/no-jquery': 'error',
        'ember/require-tagless-components': 'error',
      },
      useEslintrc: false,
    });
  }

  get rootPath(): string {
    return this.args.path;
  }

  async run(): Promise<OctaneMigrationStatusTaskResult> {
    this.report = this.esLintEngine.executeOnFiles([`${this.rootPath}/+(app|addon)/**/*.js`]);
    let result = new OctaneMigrationStatusTaskResult(this.meta, this.report);

    return result;
  }
}
