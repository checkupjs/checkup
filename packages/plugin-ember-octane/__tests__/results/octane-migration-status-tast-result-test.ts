import { MigrationType } from '../../src/results/octane-migration-status-task-result';
import { OctaneMigrationStatusTask } from '../../src/tasks';
import { OctaneMigrationStatusTaskResult } from '../../src/results';
import { stdout } from '@checkup/test-helpers';

describe('octane-migration-status-task-result', () => {
  describe('console output', () => {
    test('simple console output', async () => {
      let sampleESLintReport = require('../__fixtures__/sample-octane-eslint-report.json');
      let sampleTemplateLintReport = require('../__fixtures__/sample-octane-template-lint-report.json');
      let task = new OctaneMigrationStatusTask({});
      let taskResult = new OctaneMigrationStatusTaskResult(
        task.meta,
        sampleESLintReport,
        sampleTemplateLintReport
      );

      taskResult.stdout();

      expect(stdout()).toMatchSnapshot();
    });
  });

  describe('JSON output', () => {
    test('it should have basic JSON results', () => {
      let sampleESLintReport = require('../__fixtures__/sample-octane-eslint-report.json');
      let sampleTemplateLintReport = require('../__fixtures__/sample-octane-template-lint-report.json');
      let task = new OctaneMigrationStatusTask({});
      let taskResult = new OctaneMigrationStatusTaskResult(
        task.meta,
        sampleESLintReport,
        sampleTemplateLintReport
      );

      let jsonResults = taskResult.json();
      let { totalViolations, migrationTasks } = jsonResults.result.esLint;
      let nativeClassesMigrationInfo = migrationTasks[MigrationType.NativeClasses];

      expect(totalViolations).toBe(17);
      expect(nativeClassesMigrationInfo).toBeDefined();
      expect(nativeClassesMigrationInfo.name).toBe('Native Class');
    });

    test('it should output detailed completion data', () => {
      let sampleESLintReport = require('../__fixtures__/sample-octane-eslint-report.json');
      let sampleTemplateLintReport = require('../__fixtures__/sample-octane-template-lint-report.json');
      let task = new OctaneMigrationStatusTask({});
      let taskResult = new OctaneMigrationStatusTaskResult(
        task.meta,
        sampleESLintReport,
        sampleTemplateLintReport
      );

      let jsonResults = taskResult.json();
      let { migrationTasks } = jsonResults.result.esLint;
      let nativeClassesMigrationInfo = migrationTasks[MigrationType.NativeClasses];

      expect(nativeClassesMigrationInfo).toBeDefined();
      expect(nativeClassesMigrationInfo.completionInfo.total).toBe(11);
      expect(nativeClassesMigrationInfo.completionInfo.completed).toBe(1);
      expect(nativeClassesMigrationInfo.completionInfo.percentage).toBe('9.09');

      let taglessComponentsMigrationInfo = migrationTasks[MigrationType.TaglessComponents];

      expect(taglessComponentsMigrationInfo).toBeDefined();
      expect(taglessComponentsMigrationInfo.completionInfo.total).toBe(6);
      expect(taglessComponentsMigrationInfo.completionInfo.completed).toBe(5);
      expect(taglessComponentsMigrationInfo.completionInfo.percentage).toBe('83.33');

      let glimmerComponentMigrationInfo = migrationTasks[MigrationType.GlimmerComponents];

      expect(glimmerComponentMigrationInfo).toBeDefined();
      expect(glimmerComponentMigrationInfo.completionInfo.total).toBe(6);
      expect(glimmerComponentMigrationInfo.completionInfo.completed).toBe(4);
      expect(glimmerComponentMigrationInfo.completionInfo.percentage).toBe('66.67');

      let trackedPropertiesMigrationInfo = migrationTasks[MigrationType.TrackedProperties];

      expect(trackedPropertiesMigrationInfo).toBeDefined();
      expect(trackedPropertiesMigrationInfo.completionInfo.total).toBe(6);
      expect(trackedPropertiesMigrationInfo.completionInfo.completed).toBe(5);
      expect(trackedPropertiesMigrationInfo.completionInfo.percentage).toBe('83.33');
    });
  });
});
