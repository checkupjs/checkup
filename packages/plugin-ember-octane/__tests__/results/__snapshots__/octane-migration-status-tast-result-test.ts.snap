// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`octane-migration-status-task-result console output simple console output 1`] = `
"=== Octane Migration Status

Handlebars Octane Violations: 0
JavaScript Octane Violations: 17

Migration Task     Completion Percentage 
Native Class       9.09%                 
Tagless Component  83.33%                
Glimmer Component  66.67%                
Tracked Properties 83.33%                

"
`;
